src/Voron/Data/CompactTrees/CompactTree.Dictionaries.cs

```diff
@@ -0,0 +1,264 @@
+            // all the keys for each branch node that we encounter on the way down. We are just returning a
+            // single key from the leaf nodes because they tend to cluster too much and bias the sample. 
+
+            ref var cState = ref _cursor;
+            ref var state = ref cState._stk[cState._pos];
+
+            Span<byte> keySpan = Span<byte>.Empty;
+
+            // While we haven't seen this page, we can process it. 
+            while (state.Header->PageFlags.HasFlag(CompactPageFlags.Branch))
+            {
+                // If we have seen this page before, we just continue down the path. 
+                if (_pagesSeen.Contains(state.Page.PageNumber))
+                {
+                    var next = GetValue(ref state, _generator.Next(state.Header->NumberOfEntries));
+                    _tree.PushPage(next, ref cState);
+
+                    state = ref cState._stk[cState._pos];
+                }
+                else
+                {
+                    // We havent seen it, so we are going to be processing it. 
+
+                    // If the current index is already out of bounds, we are going to be marking this page as seen and move
+                    // to the next random page down the path.
+                    if (_currentIdx >= state.Header->NumberOfEntries)
+                    {
+                        _pagesSeen.Add(state.Page.PageNumber);
+                        _currentIdx = 0;
+
+                        // Push the next random page.
+                        var next = GetValue(ref state, _generator.Next(state.Header->NumberOfEntries));
+                        _tree.PushPage(next, ref cState);
+                        state = ref cState._stk[cState._pos];
+                    }
+                    else
+                    {
+                        GetEntry(_tree, state.Page, state.EntriesOffsets[_currentIdx], out keySpan, out var value);
+                        _currentIdx++;
+                        goto Success;
+                    }
+                }
+            }
+
+            // If the current page is a leaf we are going to retrieve a single random key and setup everything
+            // to start the random process of keys generation from the root again on the next call. 
+            if (state.Header->PageFlags.HasFlag(CompactPageFlags.Leaf))
+            {
+                int randomEntry = _generator.Next(state.Header->NumberOfEntries);
+                GetEntry(_tree, state.Page, state.EntriesOffsets[randomEntry], out keySpan, out var value);
+
+                // Move the cursor to the root.                 
+                _cursor._pos = -1;
+                _cursor._len = 0;
+                _tree.PushPage(_tree._state.RootPage, ref _cursor);
+                _currentIdx = 0;
+            }
+
+            Debug.Assert(keySpan.Length != 0, "If no key has been retrieved, there is an error in the algorithm.");
```

Comment:
That must not happen, it signals an catastrophic error that would have never ever happen in the code. Probably it should be included in a `Validation` build, but there is no reason why we want to have that check in production. It will blow up eventually up in the chain, its not that you wont realize, this is more to pinpoint the issue while debugging.